grouper/graph.py

```diff
@@ -60,46 +60,72 @@ class NoSuchGroup(Exception):
+
     Attributes:
         lock: Read lock on the data
-        update_lock: Write lock on the data
+        checkpoint: Revision of Grouper data
+        checkpoint_time: Last update time of Grouper data
         users: Names of all enabled users
         groups: Names of all enabled groups
         permissions: Names of all enabled permissions
-        checkpoint: Revision of Grouper data
-        checkpoint_time: Last update time of Grouper data
         user_metadata: Full information about each user
-        group_metadata: Full information about each group
-        group_service_accounts: Service accounts owned by groups
-        permission_metadata: Permission grant information for users
-        service_account_permission_grants: Permission grant information for service accounts
-        permission_tuples: Metadata for all enabled permissions
-        group_tuples: Metadata for all enabled groups
-        disabled_group_tuples: Metadata for all disabled groups
+        permission_grants: Permission grant information for users
     """
 
     def __init__(self):
         # type: () -> None
-        self.logger = logging.getLogger(__name__)
+        self._logger = logging.getLogger(__name__)
+
+        # lock is a read lock to ensure consistency while iterating through data.  update_lock is
+        # the write lock, held to prevent two updates from the database from running at the same
+        # time.  lock has to be public for now because some API code takes the lock and looks at
+        # data elements directly.  :(
+        self.lock = RLock()
+        self._update_lock = RLock()
+
+        # Initialized by update_from_db.
         self._graph = DiGraph()
         self._rgraph = DiGraph()
-        self.lock = RLock()
-        self.update_lock = RLock()
-        self.users = set()  # type: Set[str]
-        self.groups = set()  # type: Set[str]
-        self.permissions = set()  # type: Set[str]
+
+        # The last update sequence number and timestamp of the database underlying the graph.
         self.checkpoint = 0
         self.checkpoint_time = 0
+
+        # Collection of all groups and permissions.
+        self._groups = {}  # type: Dict[str, Group]
+        self._permissions = {}  # type: Dict[str, Permission]
+
+        # Collection of all users and their data.  For now, this is represented as a dict rather
+        # than as a data transfer object.  Users have a lot of structure, so require a more
+        # complicated object, which hasn't been written yet.
         self.user_metadata = {}  # type: Dict[str, Dict[str, Any]]
-        self.group_metadata = {}  # type: Dict[str, Dict[str, Any]]
-        self.group_service_accounts = {}  # type: Dict[str, List[str]]
+
+        # Collection of all disabled groups.
+        self._disabled_groups = set()  # type: Set[Group]
+
+        # Map of groups to their permission grants.
         self.permission_grants = {}  # type: Dict[str, List[GroupPermissionGrant]]
```

Comment:
Should this be made private as well? Doesn't seem to be used directly anywhere else, unless I'm missing something

Reply:
It's used in tests, so I was deferring cleanup of that until I could figure out a better way to implement those tests. I've added that to my to-do list for cleanups.